server/api.js

```diff
@@ -60,6 +60,57 @@ const cantPlay = (user, tourney) =>
     "Mapper",
   ]);
 
+const canEditWarmup = async (user, playerNo, match) => {
+  async function isCaptainOf(playerName, teamName, tourney) {
+    const team = await Team.findOne({ name: teamName, tourney: tourney }).populate('players');
+    if (!team || !team.players || !team.players[0]) return false;
+    return team.players[0].username === playerName;
+  }
+
+  // Admin can always edit warmup
+  if (user.isAdmin) return true;
+
+  // Players can't edit if the match is in less than 1 hour
+  if (match.time.getTime() - Date.now() < 3600000) return false;
+
+  const tourney = await Tournament.findOne({ code: match.tourney });
+  if (
+    tourney.teams &&
+    await isCaptainOf(
+      user.username,
+      match[`player${playerNo}`],
+      tourney.code)) {
+    // User is the captain of the team
+    return true;
+  } else if (user.username === match[`player${playerNo}`]) {
+    // User is the player
+    return true;
+  }
+
+  return false;
+}
+
+const parseWarmup = async (warmup) => {
+  if (!warmup) {
+    return "";
```

Comment:
not very important edge case: this means that if the user submits the empty string for warmup, it'll still stay "warmup submitted successfully"

Reply:
Fixed in ca39bbcc048976a7099528807c3595125a41012a